### DISCUSSION TITLE
Cors에러란 무엇일까요

### DISCUSSION THREAD

**Discussion by skyhanull** [wanted-9th-3team/frontend-study#10] [프론트 기술 질문 공간]
Cors에러란 무엇일까요??? SOP와 같이 설명해주세요

**Comment by michoball** [comment]
교차 출처 리소스 공유 에러를 cors 에러라고 합니다. 
옛날 방식의 웹에는 서버와 프론트 단이 분리되어있지 않았습니다. 
그렇기 때문에 하나의 동일한 출처를 가졌는데요. 

다른 출처의 리소스를 이용한다는 것은 보안상 위협이 될 수 있는 일이기 때문에 
동일한 출처에서만 리소스를 공유할 수 있도록 보안상 방어선을 만들었는데 그게 sop 입니다.

 하지만 하나의 서비스에서도 서버와 프론트의 배포를 분리해두는 현대에는 동일 출처가 아닌 서버로부터 데이터를 받아올 수 없었고 교차 출처에서 리소스를 받아오려고 시도하는 것이기에 cors 에러가 났다고 말합니다. 

개발을 하면서 위 cors에러가 나면  개발할때는 보통 axios를 쓸 때 withCredential : true 옵션을 주고 package.json에 proxy 값으로 서버 주소를 넣어 둡니다.

**Comment by WooGie911** [comment]
Cross-Origin Resource Sharing의 약자로 보안적인 이유로 브라우저에서는 Cross-Origin http 요청들을 제한합니다. 

cross-origin이란 
1. 프로토콜이 다른경우 (http, https) 
2. 도메인이 다른경우 (domain.com, other-domain.com) 
3. 포트번호가 다른경우 (8080포트, 3000포트) 
중 한가지라도 다른 경우를 의미합니다. 

허용 방법으로는 Access-Control-Allow-Origin으로 서버측 응답에서 접근 권한을 주는 헤더를 추가하여 해결하는 방법, CORS모듈을 사용하는 방법, 그리고 webpack-dev-server proxy 기능을 이용하여 리액트 개발 환경에서 중간 프록시 서버를 설정하여 서버와 같은 도메인으로 요청하여 CORS를 허용하게 하는 방법이 있습니다.

출처(Origin)의 동일함은 두 URL의 구성 요소 중 Protocol(Scheme), Host, Port 이 3가지만 동일 하다면 동일 출처로 판단하는 것으로 알고 있습니다.
